src/Microsoft.AspNet.Mvc.Core/Formatters/OutputFormatter.cs

```diff
@@ -215,16 +214,29 @@ public virtual void WriteResponseHeaders(OutputFormatterContext context)
             var selectedEncoding = SelectCharacterEncoding(context);
             if (selectedEncoding != null)
             {
-                context.SelectedEncoding = selectedEncoding;
-
                 // Override the content type value even if one already existed.
-                selectedMediaType.Charset = selectedEncoding.WebName;
+                selectedMediaType.Encoding = selectedEncoding;
```

Comment:
This is fine but double-check no later code reconfirms `SupportedMediaTypes.Any(supportedType => supportedType.IsSubsetOf(context.ContentType))` or similar. You're making `ContentType` more specific than it was.